## Pre-fix investigation (the "two of everything")

**Two hash primitives**
- `user_story_tests._story_content_hash` → `sha256(_normalized_story_for_hash(story))` — story prose only, strips `pdd-story-prompts`/`pdd-story-dev-units` comments + incidental whitespace. **Metadata-insensitive.**
- `story_test_generation.story_bundle_hash` → (before) `sha256(RAW story + "\n\n" + RAW contract)`. Included metadata comments → **metadata-sensitive** (Bug 2).

**Two evaluators**
- `_classify_story_from_markers` (full gate, `evaluate_stories`): accepts marker `story_hash=` kwarg OR module constant, over content ∪ bundle. Permissive/correct.
- `evaluate_story_regression` (wired to `pdd checkup coverage --story-regression-gate`): compared **only** bundle hash vs **only** `STORY_HASH`-constant regex — ignored kwargs (Bug 1).

**Two AST scanners**
- Gate `_module_string_constants`/`_string_value` resolve module constants.
- `story_regression._literal_story_id` accepted only `ast.Constant` → static fallback missed `story_id=CONST` (Bug 3).

**Committed-hash audit** — the seed tests already mixed forms: `pdd_sync/fix/change` recorded the **content** hash, while `pdd_update/connect/test/generate/checkup/bug` recorded the **old bundle** hash, and `top_flows pdd_generate` was already stale (`3660…` matched neither). Normalizing the bundle hash refreshes the bundle-form constants (content-form ones are unchanged and stay accepted after Bug 1).

## Red → green

Red (4 new tests fail on `HEAD~1`):
```
FAILED test_story_regression_gate.py::test_evaluate_story_regression_accepts_recorded_content_hash
FAILED test_story_regression_gate.py::test_evaluate_story_regression_stale_on_bogus_marker_kwarg_hash
FAILED test_story_regression_gate.py::test_metadata_only_relink_does_not_flip_bundle_hash
FAILED test_story_regression.py::TestStaticFallback::test_constant_story_id_markers_survive_empty_pytest_collection
```
Green after fix: `test_story_regression_gate.py`, `test_story_regression.py`, `test_story_test_generation.py`, `test_story_test_generator.py` → **87 passed**.

## E2E evidence
Wired evaluator on a story+contract (so bundle ≠ content) with a `pdd-story-prompts` comment:
```
A content_hash != bundle_hash: 1ab132798b7397de vs 96b4714503e40dcc
A (content-hash recorded)  -> story-regression-present   (was false-STALE)
metadata relink bundle stable: True
metadata-relink            -> story-regression-present   (no false stale)
B (bogus kwarg "0000…")    -> story-regression-stale     (was false-PASS)
```
`pdd checkup coverage --story-regression-gate warn` runs clean (the shown `story-regression-missing` entries are stories with no seed test — pre-existing gaps, unrelated). `evaluate_stories` over the repo: **0 stale**, 10 seed stories ok. `pytest -m story`: **15 passed**.

## How the evaluators are unified
Shared helper `_acceptable_story_hashes(story_path)` (content ∪ bundle) is called by BOTH `_classify_story_from_markers` and `evaluate_story_regression`; the latter now collects recorded hashes with the gate's own `_markers_in_source` AST scan (via `_recorded_story_hashes`), so kwarg/constant resolution is single-sourced. Bundle normalization lives in `story_test_generation._normalized_bundle`; the generated text-pin's inline `_bundle_hash()` imports `story_bundle_hash` so it cannot drift from the recorded constant.

`git log origin/codex/issue-1816-story-regression-epic..HEAD --oneline`:
```
1568f65aa fix(story): unify story-hash freshness semantics across gate evaluators (pdd#1889)
b95fc015f test(story): pin unified story-hash freshness semantics (pdd#1889)
```

--- a/pdd/story_regression_gate.py
+++ b/pdd/story_regression_gate.py
@@ -808,6 +808,7 @@ def evaluate_story_regression(
         )
 
     recorded: dict[str, str] = {}
+    all_recorded: set = set()
     for nodeid in tests:
         test_path = _test_file_for_nodeid(nodeid, tests_dir)
         if test_path is None:
@@ -815,6 +816,10 @@ def evaluate_story_regression(
         found = _recorded_story_hashes(test_path, sid)
         if found:
             recorded[nodeid] = found[0]
+            # Freshness is decided over EVERY recorded hash, not just the first,
+            # so a stale marker listed before a fresh one in the same file cannot
+            # shadow it -- matching the full gate's any-fresh-wins rule (#1889).
+            all_recorded.update(found)
 
     if not recorded:
         # Coverage uses this lightweight evaluator to preserve #1699
@@ -832,7 +837,7 @@ def evaluate_story_regression(
     # Same acceptance as the full gate: content hash UNION bundle hash. A
     # recorded hash matching either form is fresh; otherwise the story changed.
     acceptable = _acceptable_story_hashes(story_path)
-    if set(recorded.values()) & acceptable:
+    if all_recorded & acceptable:
         return StoryRegressionEvaluation(
             story_id=sid,
             status=STATUS_PASSING,

--- a/tests/test_coverage_contracts.py
+++ b/tests/test_coverage_contracts.py
@@ -1225,7 +1225,7 @@ class TestStoryRegressionDimension:
                 "has_regression_test": True,
                 "status": "story-regression-present",
                 "tests": ["test_foo.py::test_foo_flow"],
-                "story_hash": "8ca87cfe7267abcb",
+                "story_hash": "95d7a1ca9e45f480",
             }
         ]
         assert "regression_warnings" in d

--- a/tests/test_story_regression_gate.py
+++ b/tests/test_story_regression_gate.py
@@ -568,6 +568,34 @@ def test_evaluate_story_regression_accepts_recorded_content_hash(tmp_path: Path)
     assert ev.status == STATUS_PASSING
 
 
+def test_evaluate_story_regression_fresh_wins_over_earlier_stale_in_same_file(
+    tmp_path: Path,
+):
+    """A file with two markers for the same story -- a STALE one listed first,
+    a FRESH one second -- must be reported 'present' (any-fresh-wins), matching
+    the full gate. Regression guard for collapsing the recorded-hash list to
+    ``found[0]``, which would re-introduce exactly the evaluator divergence this
+    change eliminates (pdd#1889)."""
+    from pdd.story_regression import build_story_map
+    from pdd.story_regression_gate import STATUS_PASSING, evaluate_story_regression
+
+    stories = tmp_path / "user_stories"
+    tests = tmp_path / "tests"
+    story = _write(stories / "story__refund.md", FRESH_STORY)
+    content_hash = _story_content_hash(FRESH_STORY)
+    # Stale marker FIRST (source order drives AST scan order), fresh SECOND.
+    _write(
+        tests / "test_refund.py",
+        _test_module(
+            _marked("refund", "0000000000000000", "test_refund_stale"),
+            _marked("refund", content_hash, "test_refund_fresh"),
+        ),
+    )
+    smap = build_story_map(tests)
+    ev = evaluate_story_regression(story, tests_dir=tests, story_map=smap)
+    assert ev.status == STATUS_PASSING
+
+
 def test_evaluate_story_regression_stale_on_bogus_marker_kwarg_hash(tmp_path: Path):
     """Direction B (false PASS): a test recording only a bogus ``story_hash=``
     marker kwarg (no module constant) is genuinely stale and must be reported

